### PR TITLE
makes shutters buildable

### PR DESCRIPTION
**Pull request by triplezeta**
## About The Pull Request

partial port of https://github.com/tgstation/tgstation/pull/56486

makes shutters and blast doors buildable. use plasteel sheetcrafting to build em. Decon steps are also included.

## Why It's Good For The Game

requested for awhile

## Changelog
:cl:
add: you can build shutters now!
/:cl:

### CODE REVIEW

**Comment by triplezeta** [comment]
note thsi pr is kind of weird rn, i left a lot of update_icons procs in there despite there being no icons to change

**Comment by triplezeta** [comment]
aside from that the PR works perfectly well after light testing. With the exception of whatever the hell is going on here 
image

**Comment by triplezeta** [comment]
the blast door needs to be open in order to do anything to it and it leaves behind the assembly, which needs a welding and wrenching step to dislodge.

I can also make it automatically shut itself while power is on if you try prying out electronics, along with the shock, and uparmor the assemblies.

File: code/game/machinery/doors/poddoor.dm [hunk 1]
```diff
@@ -15,8 +15,57 @@
 	resistance_flags = FIRE_PROOF
 	damage_deflection = 70
 	poddoor = TRUE
+	assemblytype = /obj/structure/poddoor_assembly
 	var/open_sound = 'sound/machines/blastdoor.ogg'
 	var/close_sound = 'sound/machines/blastdoor.ogg'
+	var/ertblast = FALSE //If this is true the blast door cannot be deconstructed
```
Comment (Fikou) [inline code/game/machinery/doors/poddoor.dm]: use the nodeconstruct flag we already have?

File: code/game/machinery/doors/poddoor.dm [hunk 2]
```diff
@@ -15,8 +15,57 @@
...
+			var/change_id = input("Set [src]'s ID. It must be a number between 1 and 100.", "ID", id) as num|null
+			if(change_id)
+				id = clamp(round(change_id, 1), 1, 100)
+				to_chat(user, "<span class='notice'>You change the ID to [id].</span>")
+
+		if(W.tool_behaviour == TOOL_CROWBAR)
+			to_chat(user, "<span class='notice'>You start to remove the airlock electronics.</span>")
+			if(!(machine_stat & NOPOWER))
+				electrocute_mob(user, get_area(src), src, 1, TRUE) //fuck this fella
+			if(W.use_tool(src, user, 100, volume=50))
```
Comment (Fikou) [inline code/game/machinery/doors/poddoor.dm]: use SECONDS defines

File: code/game/machinery/doors/poddoor.dm [hunk 3]
```diff
@@ -15,8 +15,57 @@
...
+/obj/machinery/door/poddoor/examine(mob/user)
+	. = ..()
+	. += "<span class='notice'>The maintenance panel is [panel_open ? "opened" : "closed"].</span>"
+	if(panel_open)
+		. += "<span class='notice'>The <b>airlock electronics</b> are exposed and could be <i>pried out</i>."
+
+/obj/machinery/door/poddoor/deconstruct(disassembled = TRUE, mob/user)
+	if(!(flags_1 & NODECONSTRUCT_1))
+		var/obj/structure/poddoor_assembly/A
+		A = new assemblytype(loc)
```
Comment (Fikou) [inline code/game/machinery/doors/poddoor.dm]: you can merge these into one line, also use a better var name than A please

File: code/game/objects/structures/poddoor_assembly.dm [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+		if(AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS)
+			. += "<span class='notice'>The maintenance panel is <b>wired</b>, but the circuit slot is <i>empty</i>.</span>"
+		if(AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER)
+			. += "<span class='notice'>The circuit is <b>connected loosely</b> to its slot.</span>"
+	. += "<span class='notice'>The assembly [welded ? "is firmly <b>welded</b> into place" : "needs to be <i>welded</i> to the floor in order to complete it"].</span>"
+	. += "<span class='notice'>There is a small placard on the assembly[doorname].</span>"
+
+/obj/structure/poddoor_assembly/attackby(obj/item/W, mob/user, params)
+	if(istype(W, /obj/item/pen))
+		var/t = stripped_input(user, "Enter the name for the assembly.", name, created_name,MAX_NAME_LEN)
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: use a better variable name than t

File: code/game/objects/structures/poddoor_assembly.dm [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+		created_name = t
+
+	else if(W.tool_behaviour == TOOL_WELDER)
+		if(!W.tool_start_check(user, amount=0))
+			return
+
+		else if(!anchored)
+			user.visible_message("<span class='warning'>[user] cuts apart [src].</span>", \
+								"<span class='notice'>You start to slice apart [src]...</span>")
+			if(W.use_tool(src, user, 40, volume=50))
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: use SECONDS defines

File: code/game/objects/structures/poddoor_assembly.dm [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+/obj/structure/poddoor_assembly/attackby(obj/item/W, mob/user, params)
+	if(istype(W, /obj/item/pen))
+		var/t = stripped_input(user, "Enter the name for the assembly.", name, created_name,MAX_NAME_LEN)
+		if(!t || (!in_range(src, usr) && loc != usr) )
+			return
+		created_name = t
+
+	else if(W.tool_behaviour == TOOL_WELDER)
+		if(!W.tool_start_check(user, amount=0))
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: does check for 0 work?

File: code/game/objects/structures/poddoor_assembly.dm [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+		else if(!anchored)
+			user.visible_message("<span class='warning'>[user] cuts apart [src].</span>", \
+								"<span class='notice'>You start to slice apart [src]...</span>")
+			if(W.use_tool(src, user, 40, volume=50))
+				to_chat(user, "<span class='notice'>You disassemble [src].</span>")
+				deconstruct(TRUE)
+		else if(!welded)
+			user.visible_message("<span class='warning'>[user] welds [src].</span>", \
+								"<span class='notice'>You start to weld [src] to the floor...</span>")
+			if(W.use_tool(src, user, 40, volume=50))
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: SECONDS defines

File: code/game/objects/structures/poddoor_assembly.dm [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+				if(!anchored || welded)
+					return
+				to_chat(user, "<span class='notice'>You weld [src] to the floor.</span>")
+				welded = TRUE
+		else
+			user.visible_message("<span class='warning'>[user] welds [src].</span>", \
+								"<span class='notice'>You start to weld [src] free from the floor...</span>")
+			if(W.use_tool(src, user, 40, volume=50))
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: ditto

File: code/game/objects/structures/poddoor_assembly.dm [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+				if(!welded)
+					return
+				to_chat(user, "<span class='notice'>You weld [src] from the floor.</span>")
+				welded = FALSE
+
+	else if(W.tool_behaviour == TOOL_WRENCH)
+		if(!anchored )
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]:
```suggestion
		if(!anchored)
```

File: code/game/objects/structures/poddoor_assembly.dm [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+			var/door_check = 1
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: use TRUE/FALSE

File: code/game/objects/structures/poddoor_assembly.dm [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+				return
+
+			if(door_check)
+				user.visible_message(
+					"<span class='notice'>[user] secures [src] to the floor.</span>",
+					"<span class='notice'>You start to secure [src] to the floor...</span>",
+					"<span class='hear'>You hear wrenching.</span>"
+				)
+
+				if(W.use_tool(src, user, 40, volume=100))
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: again

File: code/game/objects/structures/poddoor_assembly.dm [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+			else
+				to_chat(user, "There is another [base_name] here!")
+
+		else
+			user.visible_message(
+				"<span class='notice'>[user] unsecures [src] from the floor.</span>",
+				"<span class='notice'>You start to free [src] from the floor...</span>",
+				"<span class='hear'>You hear wrenching.</span>"
+			)
+			if(W.use_tool(src, user, 40, volume=100))
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: yeah

File: code/game/objects/structures/poddoor_assembly.dm [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+				if(!anchored)
+					return
+				to_chat(user, "<span class='notice'>You unsecure [src].</span>")
+				name = "airlock assembly"
+				set_anchored(FALSE)
+
+	else if(istype(W, /obj/item/stack/cable_coil) && state == AIRLOCK_ASSEMBLY_NEEDS_WIRES )
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]:
```suggestion
	else if(istype(W, /obj/item/stack/cable_coil) && state == AIRLOCK_ASSEMBLY_NEEDS_WIRES)
```

File: code/game/objects/structures/poddoor_assembly.dm [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+
+		user.visible_message("<span class='notice'>[user] wires [src].</span>", \
+							"<span class='notice'>You start to wire [src]...</span>")
+		if(W.use_tool(src, user, 40, amount=1))
+			if(state != AIRLOCK_ASSEMBLY_NEEDS_WIRES)
+				return
+			state = AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS
+			to_chat(user, "<span class='notice'>You wire [src].</span>")
+
+	else if((W.tool_behaviour == TOOL_WIRECUTTER) && state == AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS )
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]:
```suggestion
	else if((W.tool_behaviour == TOOL_WIRECUTTER) && state == AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS)
```

File: code/game/objects/structures/poddoor_assembly.dm [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+		user.visible_message("<span class='notice'>[user] cuts the wires from [src].</span>", \
+							"<span class='notice'>You start to cut the wires from [src]...</span>")
+
+		if(W.use_tool(src, user, 40, volume=100))
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: SECONDS

File: code/game/objects/structures/poddoor_assembly.dm [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+				return
+			to_chat(user, "<span class='notice'>You cut the wires from the [src].</span>")
+			new/obj/item/stack/cable_coil(get_turf(user), 1)
+			state = AIRLOCK_ASSEMBLY_NEEDS_WIRES
+
+	else if(istype(W, /obj/item/electronics/airlock) && state == AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS )
+		W.play_tool_sound(src, 100)
+		user.visible_message("<span class='notice'>[user] installs the electronics into [src].</span>", \
+							"<span class='notice'>You start to install electronics into [src]...</span>")
+		if(do_after(user, 40, target = src))
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: SECONDS

File: code/game/objects/structures/poddoor_assembly.dm [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+			if( state != AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS )
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]:
```suggestion
			if(state != AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS)
```

File: code/game/objects/structures/poddoor_assembly.dm [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+	else if((W.tool_behaviour == TOOL_CROWBAR) && state == AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER )
+		user.visible_message("<span class='notice'>[user] removes the electronics from [src].</span>", \
+							"<span class='notice'>You start to remove electronics from [src]...</span>")
+
+		if(W.use_tool(src, user, 40, volume=100))
+			if(state != AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER)
+				return
+			to_chat(user, "<span class='notice'>You remove the electronics.</span>")
+			state = AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS
+			var/obj/item/electronics/airlock/ae
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: better variable name

File: code/game/objects/structures/poddoor_assembly.dm [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+			if (!electronics)
+				ae = new/obj/item/electronics/airlock( loc )
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]:
```suggestion
				ae = new /obj/item/electronics/airlock(loc)
```

File: code/game/objects/structures/poddoor_assembly.dm [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+			else
+				ae = electronics
+				electronics = null
+				ae.forceMove(src.loc)
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: don't do src.var

File: code/game/objects/structures/poddoor_assembly.dm [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+
+	else if((W.tool_behaviour == TOOL_SCREWDRIVER) && state == AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER && welded )
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]:
```suggestion
	else if((W.tool_behaviour == TOOL_SCREWDRIVER) && state == AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER && welded)
```

File: code/game/objects/structures/poddoor_assembly.dm [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+			if( state != AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS )
+				return
+			if(!user.transferItemToLoc(W, src))
+				return
+
+			to_chat(user, "<span class='notice'>You install the electronics.</span>")
+			state = AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER
+			electronics = W
+
+	else if((W.tool_behaviour == TOOL_CROWBAR) && state == AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER )
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]:
```suggestion
	else if((W.tool_behaviour == TOOL_CROWBAR) && state == AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER)
```

File: code/game/objects/structures/poddoor_assembly.dm [hunk 23]
```diff
@@ -0,0 +1,220 @@
...
+							"<span class='notice'>You start to cut the wires from [src]...</span>")
+
+		if(W.use_tool(src, user, 40, volume=100))
+			if(state != AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS)
+				return
+			to_chat(user, "<span class='notice'>You cut the wires from the [src].</span>")
+			new/obj/item/stack/cable_coil(get_turf(user), 1)
+			state = AIRLOCK_ASSEMBLY_NEEDS_WIRES
+
+	else if(istype(W, /obj/item/electronics/airlock) && state == AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS )
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]:
```suggestion
	else if(istype(W, /obj/item/electronics/airlock) && state == AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS)
```

File: code/game/objects/structures/poddoor_assembly.dm [hunk 24]
```diff
@@ -0,0 +1,220 @@
...
+
+			to_chat(user, "<span class='notice'>You install the electronics.</span>")
+			state = AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER
+			electronics = W
+
+	else if((W.tool_behaviour == TOOL_CROWBAR) && state == AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER )
+		user.visible_message("<span class='notice'>[user] removes the electronics from [src].</span>", \
+							"<span class='notice'>You start to remove electronics from [src]...</span>")
+
+		if(W.use_tool(src, user, 40, volume=100))
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: seconds

File: code/game/objects/structures/poddoor_assembly.dm [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+				ae = electronics
+				electronics = null
+				ae.forceMove(src.loc)
+
+	else if((W.tool_behaviour == TOOL_SCREWDRIVER) && state == AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER && welded )
+		user.visible_message(
+			"<span class='notice'>[user] finishes [src].</span>",
+			"<span class='notice'>You start finishing [src]...</span>"
+		)
+		if(W.use_tool(src, user, 40, volume=100))
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: and final seconds

File: code/game/objects/structures/poddoor_assembly.dm [hunk 26]
```diff
@@ -0,0 +1,220 @@
...
+			if(!welded)
+				return
+			if(loc && state == AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER)
+				to_chat(user, "<span class='notice'>You finish [src].</span>")
+				var/obj/machinery/door/poddoor/door
+				door = new poddoor_type(loc)
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: you can merge these

File: code/game/objects/structures/poddoor_assembly.dm [hunk 27]
```diff
@@ -0,0 +1,220 @@
...
+			else
+				. +=  "<span class='notice'>The anchoring bolts are <i>unwrenched</i>, and the assembly could be <i>welded apart</i>.</span>"
+		if(AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS)
+			. += "<span class='notice'>The maintenance panel is <b>wired</b>, but the circuit slot is <i>empty</i>.</span>"
+		if(AIRLOCK_ASSEMBLY_NEEDS_SCREWDRIVER)
+			. += "<span class='notice'>The circuit is <b>connected loosely</b> to its slot.</span>"
+	. += "<span class='notice'>The assembly [welded ? "is firmly <b>welded</b> into place" : "needs to be <i>welded</i> to the floor in order to complete it"].</span>"
+	. += "<span class='notice'>There is a small placard on the assembly[doorname].</span>"
+
+/obj/structure/poddoor_assembly/attackby(obj/item/W, mob/user, params)
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]: i would appreciate a better variable name than W

**Review by Fikou — CHANGES_REQUESTED** [review]

**Comment by triplezeta** [comment]
addressed reviews + blast doors need to be open and depowered in order to be deconstructed

File: code/game/objects/structures/poddoor_assembly.dm (hunk 6 above)
Comment (triplezeta) [reply code/game/objects/structures/poddoor_assembly.dm]: from what i can tell it sets the amount of fuel used to zero at the start of welding. Works as intended in testing, at least.

File: code/game/objects/items/stacks/sheets/sheet_types.dm [hunk 28]
```diff
@@ -166,7 +166,9 @@ GLOBAL_LIST_INIT(plasteel_recipes, list ( \
 		new/datum/stack_recipe("nanoweave (beige)", /obj/item/stack/tile/carpet/nanoweave/beige, 1, 4, 20), \
 		new/datum/stack_recipe("nanoweave (purple)", /obj/item/stack/tile/carpet/nanoweave/purple, 1, 4, 20), \
 		new/datum/stack_recipe("nanoweave (orange)", /obj/item/stack/tile/carpet/nanoweave/orange, 1, 4, 20), \
-	))
+	)), \
+	new/datum/stack_recipe("shutters frame", /obj/structure/poddoor_assembly/shutters, 5, time = 50, one_per_turf = 1, on_floor = 1 ), \
+	new/datum/stack_recipe("blast door frame", /obj/structure/poddoor_assembly, 15, time = 50, one_per_turf = 1, on_floor = 1 )
```
Comment (Fikou) [inline code/game/objects/items/stacks/sheets/sheet_types.dm]:
```suggestion
	new/datum/stack_recipe("shutters frame", /obj/structure/poddoor_assembly/shutters, 5, time = 50, one_per_turf = 1, on_floor = 1), \
	new/datum/stack_recipe("blast door frame", /obj/structure/poddoor_assembly, 15, time = 50, one_per_turf = 1, on_floor = 1)
```

File: code/game/objects/structures/poddoor_assembly.dm [hunk 29]
```diff
@@ -0,0 +1,219 @@
...
+
+		user.visible_message("<span class='notice'>[user] wires [src].</span>", \
+							"<span class='notice'>You start to wire [src]...</span>")
+		if(item_used.use_tool(src, user, 4 SECONDS, amount=1))
+			if(state != AIRLOCK_ASSEMBLY_NEEDS_WIRES)
+				return
+			state = AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS
+			to_chat(user, "<span class='notice'>You wire [src].</span>")
+
+	else if((item_used.tool_behaviour == TOOL_WIRECUTTER) && state == AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS)
```
Comment (Fikou) [inline code/game/objects/structures/poddoor_assembly.dm]:
```suggestion
	else if(item_used.tool_behaviour == TOOL_WIRECUTTER && state == AIRLOCK_ASSEMBLY_NEEDS_ELECTRONICS)
```
no need for these

**Review by Fikou — APPROVED** [review]

**Pull request merged**